### DISCUSSION TITLE
CSC calculation(Waiting for VASP lock to appear...)

### DISCUSSION THREAD

**Discussion by quantumable** [TRIQS/dft_tools#188] [Q&A]
Hi,

I can't do the csc calculation of the tutorial. I first did the single-shot calculation which ended well when I run the CSC calculation(`vasp_dmft -n 144 -i 15 -j 1 -p ~/triqs/install/bin/vasp_std csc.py`), it stops on the following message "Waiting for VASP lock to appear..." and runs for hours without progressing and this with or without` ICHARG = 5; NELM = 1000; NELMIN = 1000 ;IMIX=0`. I have compiled vasp with MKL and with lapack in both cases I have the same result.
how to calculate csc using vasp? 

PS: I have compiled vasp and triqs vasp the same configurations of the gcc compiler

**Comment by the-hampel** [comment]
Please share more information to get an answer. Is Vasp running at all? Did you see any specific Vasp output in the std out or the folder where you run the job? Any error messages?

Did you modify the Vasp source code according to the instructions listed here: https://triqs.github.io/dft_tools/unstable/guide/conv_vasp.html#remarks-on-the-vasp-version .

**Reply by quantumable** [reply]
Yes indeed I have modified the files (locproj.F, electron.F, fileio.F, main.F which I also attach here) as indicated here: [](https://triqs.github.io/dft_tools/unstable/guide/conv_vasp.html#remarks-on-the-vasp-version).
here is the message I have in my output slurm file: it just displays this message and runs for hours without anything happening.


```
>  VASP version not specified, setting to standard
>   Number of cores: 144
>   Number of iterations: 15
>   Number of iterations with fixed density: 1
>   VASP version: standard
>   Script name: f.py
>   Waiting for VASP lock to appear...
> mpirun: Forwarding signal 18 to job

```
The INCAR ;  the same plo.cfg than the on-shot


```
System  = NiO

ISMEAR = -5
EDIFF = 1.E-7
PREC = accurate
NCORE = 4
#NBANDS = 24
EMIN = -3
EMAX = 10
ICHARG = 5
NELM = 1000
NELMIN = 1000
IMIX=0

LMAXMIX = 6
ISYM = -1
LORBIT = 14
LOCPROJ = 1 : d : Pr
LOCPROJ = 2 : p : Pr
```

and the run script

```

> #!/bin/bash
> #SBATCH -J name
> #SBATCH -N 6
> #SBATCH --ntasks-per-node=24
> #SBATCH -n 144
> #SBATCH -t 24:00:00
> #SBATCH --mem=118000
> module load gcc/gcc-9  openmpi/openmpi-4.0.gcc9  boost/boost-1.69.0.g9 fftw/fftw3.3.8.g9 scalapack/scalapack-2.1.0.g9
> source /triqs/install/share/triqsvars.sh
> 
> vasp_dmft -n 144 -i 15 -j 1 -p ~/triqs/install/bin/vasp_std csc.py
> 

```

**Reply by the-hampel** [reply]
This seems as if Vasp is not starting at all. Can you first try to run the script on a single node. Is there also an error file? 

There should be immediately output from Vasp in the usual format. No need to wait for hours. The bash script `vasp_dmft` is just executing: 
```
stdbuf -o 0 $MPIRUN_CMD -np $NPROC "$VASP_DIR" &
```
which apparently fails. You could try to do this manually in your job file and see if Vasp can start. So just put this line in your job file and specify the variables. The default value for `$MPIRUN_CMD` is `mpirun`. Is that correct for your cluster? 

And please do not share Vasp source code files publicly! The Vasp code is copyright protected and not an open source code. Please remove it.

**Reply by quantumable** [reply]
it is explicitly marked in vasp_dmft that $MPIRUN_CMD = mpirun
I have tried on nodes that have 24 cores(`vasp_dmft -n 24 -i 15 -j 1 -p ~/triqs/install/bin/vasp_std csc.py`) and I still have the calculation that stops on: 
```
>  LDA part: xc-table for Pade appr. of Perdew
>  found WAVECAR, reading the header
>  POSCAR, INCAR and KPOINTS ok, starting setup
>  FFT: planning ...
>  reading WAVECAR
>   Waiting for VASP lock to appear...
>  the WAVECAR file was read successfully
>  charge-density read from file: NiO
>  entering main loop
>        N       E                     dE             d eps       ncg     rms          rms(c)
>   Waiting for VASP lock to disappear...
```
on the other hand when I launch it on a single core(`vasp_dmft -n 1 -i 15 -j 1 -p ~/triqs/install/bin/vasp_std csc.py`), the calculation runs immediately. however after several iteration I have another error that I do not know how to explain. I attach here the csc.py script and the output file when running on 1 core. at the end of the output file there is an error message.

[csc_nio_ouput_for_1core.zip](https://github.com/TRIQS/dft_tools/files/7426951/csc_nio_ouput_for_1core.zip)

```

> Starting run with 1 MPI rank(s) at : 2021-10-27 15:49:36.893465
> Traceback (most recent call last):
>   File "/home/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/home/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 257, in <module>
>     main()
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 254, in main
>     run_all(vasp_pid, dmft_mod.dmft_cycle, cfg_file, n_iter, n_iter_dft, vasp_version)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 139, in run_all
>     corr_energy, dft_dc = dmft_cycle()
>   File "/home/TUTO/TEST/r.py", line 123, in dmft_cycle
>     SK.symm_deg_gf(S.Sigma_iw,ish=0)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/sumk_dft.py", line 1847, in symm_deg_gf
>     helper.from_L_G_R(v.conjugate().transpose(), gf_to_symm[key], v)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs/gf/gf.py", line 673, in from_L_G_R
>     assert L.shape[1] == G.target_shape[0], "Dimension mismatch between L and G"
> AssertionError: Dimension mismatch between L and G
> --------------------------------------------------------------------------
> Primary job  terminated normally, but 1 process returned
> a non-zero exit code. Per user-direction, the job has been aborted.
> 

```

**Reply by quantumable** [reply]
please do you have any idea?

**Reply by the-hampel** [reply]
It seems that everything runs fine with 1 core. Your calculation is crashing because it diverges. Your number of Monte Carlo samples etc. is too small. Probably because you used only one core for DMFT. This leads to a change in block structure during your CSC steps, and finally to inconsistencies between the loaded Sigma from the previous DMFT iteration, making it impossible for triqs to use it. 

I don't know why it does not work for your with more than one core. But what you can do is to run Vasp and triqs with different numbers of cores (which is anyway advisable). Open the `vasp_dmft` script (you can locate it with the command `which vasp_dmft`), and change the line: 
```
stdbuf -o 0 $MPIRUN_CMD -np $NPROC "$VASP_DIR" &
```
to 
```
stdbuf -o 0 $MPIRUN_CMD -np 1 "$VASP_DIR" &
```
and see if it works now with your original script on multiple nodes.If it works try to increase the number of cores in the script.

**Reply by quantumable** [reply]
in fact by changing `stdbuf -o 0 $MPIRUN_CMD -np $NPROC "$VASP_DIR" &` `by stdbuf -o 0 $MPIRUN_CMD -np 1 "$VASP_DIR" &`  and also by increasing the parameter to `p["n_warmup_cycles"] = 20000 ;  p["n_cycles"] = 200000`  I had the same mistake, please have an idea?

```
Traceback (most recent call last):
  File "/home/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 257, in <module>
    main()
  File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 254, in main
    run_all(vasp_pid, dmft_mod.dmft_cycle, cfg_file, n_iter, n_iter_dft, vasp_version)
  File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 139, in run_all
    corr_energy, dft_dc = dmft_cycle()
  File "/home/TUTO/csc.py", line 123, in dmft_cycle
    SK.symm_deg_gf(S.Sigma_iw,ish=0)
  File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/sumk_dft.py", line 1847, in symm_deg_gf
    helper.from_L_G_R(v.conjugate().transpose(), gf_to_symm[key], v)
  File "/home/triqs/install/lib/python3.8/site-packages/triqs/gf/gf.py", line 673, in from_L_G_R
    assert L.shape[1] == G.target_shape[0], "Dimension mismatch between L and G"
AssertionError: Dimension mismatch between L and G
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 16 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

thanks

**Reply by the-hampel** [reply]
Did you delete all previous run data, so that the calculation starts from scratch? I think it might loaded the old data and run in the same problem. This is not a problem of CSC. The block structure of your impurity changes causing trouble.

**Reply by quantumable** [reply]
Thank you for this suggestion. 
1-) By changing `stdbuf -o 0 $MPIRUN_CMD -np $NPROC "$VASP_DIR" &` `by stdbuf -o 0 $MPIRUN_CMD -np 1 "$VASP_DIR" &` ;  this only works on one node ie `vasp_dmft -n 24 -i 15 -j 1 -p ~/triqs/install/bin/vasp_std csc.py` but not for example  `vasp_dmft -n 144 -i 15 -j 1 -p ~/triqs/install/bin/vasp_std csc.py`( and I still have the calculation that stops on `....Waiting for VASP lock to disappear...`)


2-) After the cacul dmft I created a new folder where I copied the files of the dmft calculation and at each refresh error the files n deleting the files and copying again from the folder dmft. I use of course the script daft et dmft+dft present on your site. I do not know where the error can come from.

```
> ```
> Traceback (most recent call last):
>   File "/home/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/home/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 257, in <module>
>     main()
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 254, in main
>     run_all(vasp_pid, dmft_mod.dmft_cycle, cfg_file, n_iter, n_iter_dft, vasp_version)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/converters/plovasp/sc_dmft.py", line 139, in run_all
>     corr_energy, dft_dc = dmft_cycle()
>   File "/home/TUTO/csc.py", line 123, in dmft_cycle
>     SK.symm_deg_gf(S.Sigma_iw,ish=0)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs_dft_tools/sumk_dft.py", line 1847, in symm_deg_gf
>     helper.from_L_G_R(v.conjugate().transpose(), gf_to_symm[key], v)
>   File "/home/triqs/install/lib/python3.8/site-packages/triqs/gf/gf.py", line 673, in from_L_G_R
>     assert L.shape[1] == G.target_shape[0], "Dimension mismatch between L and G"
> AssertionError: Dimension mismatch between L and G

```


 What parameter is responsible for this?

**Reply by quantumable** [reply]
Dear Alex,
 I have to post in the main group to know if another person has succeeded in making the cdc calculation with VASP and has an idea of how to do it?
Bests

**Reply by the-hampel** [reply]
Hi,
1) but this means that in principle your calculation seems to work. Than the problems are related to your cluster environment, and I can't help you there. For me this works as expected on our cluster. Meaning I use VASP for CSC DMFT calculation regularly. Unless you provide me a detailed error message why Vasp is not starting I cannot help you. The message that you show me `....Waiting for VASP lock to disappear...` is a regular output. Not the error message. 

2) Which files did you copy? Did you copy the h5 file called `vasp.h5` ? You need to recreate this file, because the results are stored inside the file. I am very certain that this is the problem.

**Reply by quantumable** [reply]
Dear Alex,

1-)when I do the usual Vasp calculation mpirun -np 144 Vasp_std it works fine. but does not work on multiple cores and nodes if I use the CSC calculation mode.

2-)Indeed I deleted the .h5 file and redid the single-shot dmft calculation. and restarted the calculation. and it just doesn't work.

I did install mpi4py and installed the code via the source.

**Reply by quantumable** [reply]
Dear Alex, 

1-) I solved the problem by changing the threshold parameter in the DMFT and DFT+DMFT calculation and the calculation goes to the end.

2-) however the calculation works only if I modify vasp_dmt as you suggested(`stdbuf -o 0 $MPIRUN_CMD -np $NPROC "$VASP_DIR" & by stdbuf -o 0 $MPIRUN_CMD -np 1 "$VASP_DIR" &`) and on only 1 node(of 24core in my case). do you have an idea why the CSC calculation does not work on several nodes?

3-) I also have the last version of vasp6.2.1 have you ever tried this version with triqs because on the file to modify main.F is not the same as that of version 5.4.4

4-)for my system for example which is composed of molecules + substrate and whose selected energy window (EWINDOW in plo.cfg) contains a lot of electrons (for example 2000 or more) the dmft calculation is not running. but if I reduce the window the calculation is running. the concern is that by reducing the window where the d-electron is present, I am afraid of not capturing the physics that it describes. 
```

> 
> 
> PS: in the script it is also necessary to modify 
> mpi.report("Orbital %s density: %.6f"%(sig,dm[sig][0,0]))
> mpi.report('Total charge of Gloc : %.6f'%S.G_iw.total_density())
>  
> to
> 
> mpi.report("Orbital %s density: %.6f"%(sig,dm[sig][0,0].real))
> mpi.report('Total charge of Gloc : %.6f'%S.G_iw.total_density().real)
> 

```

**Comment by the-hampel** [comment]
No I did not try with the latest version of Vasp. In principle it should work, but as the source code slightly changed I cannot give you exact patch files. 

I never tested the converter with that many electronic states in the window, and I think I am also not 100% sure why you need that. Do all of the 2000 electronic states hybridize with the target d states? If not you can start cutting them out. But again these things are physical questions, which I do not intend to help you with.

If you want to fix the Vasp CSC I really need the raw stdout and stderr files, that slurm creates to analyze any further what goes wrong. We saw that Vasp is not correctly starting in that case. I suspect that this is due to the cluster mpi configuration.
